specs/agents/logging.md

```diff
@@ -31,11 +30,115 @@ agents can treat that as a synonym for `error` or `fatal`.
+
+* **Agent**: This block is mandatory and contains basic version and build date information.
+* **Environment**: This block is optional but for supportability reasons it should be provided.
+* **Configuration**: This block mandatory and contains a minimum set of relevant configuration values.
+
+### Agent
+
+On startup, all APM agents MUST log basic information regarding their technology (language, runtime),
+version information, and build date in a format chosen by the respective agent.
+
+This SHOULD be the very first log message that is created by an agent.
+
+Example:
+
+```text
+Elastic APM .NET Agent, version: 1.19.1-preview, build date: 2022-10-27 10:55:42 UTC
+```
+
+Agents SHOULD also report when they are running in a **non-production** scenario (e.g. a debug
+or pre-release build) using the `warning` logging level.
+
+Example:
+
+```text
+This a pre-release version and not intended for use in production environments!
+```
+
+### Environment
+
+Additionally, agents SHOULD report information about their environment (e.g. host, process, runtime).
+
+| Item | Description | Example |
+| - | - | - |
+| Process ID | The Process ID in decimal format. | `83606` |
+| Process Name | The executable image name or the full path to it.  | `w3wp.exe`, `/usr/local/share/dotnet/dotnet` |
+| Command Line | The full command line used to launch this process as available to the runtime. | `/Users/acme/some_app/bin/Debug/net7.0/some_app.dll foo=bar` |
+| Operating System | OS name and version in a human-readable format. | `macOS Version 12.6.1 (build 21G217)` |
+| CPU architecture | One of the well-known values from the table below. | `arm64` |
+| Host | The (optionally fully-qualified) host name. | `MacBook-Pro.localdomain` |
+| Time zone | The local time zone in UTC-offset notation. | `UTC+0200` |
+| Runtime | Name and version of the executing runtime. | `.NET Framework 4.8.4250.0`|
+
+**CPU Architecture:**
+
+| Value | Description |
+| - | - |
+| `amd64` | AMD64 |
+| `arm32` |ARM32 |
+| `arm64` |ARM64 |
+| `ia64` | Itanium |
+| `ppc32` | 32-bit PowerPC |
+| `ppc64` | 64-bit PowerPC |
+| `s390x` | IBM z/Architecture |
+| `x86` | 32-bit x86 |
+
+### Configuration
+
+The start of the configuration block MUST be denoted as such (e.g. `Agent Configuration:`).
+
+If configuration files are used in the configuration process, their fully-qualified paths
+SHOULD be logged.
+
+Configuration item names MUST be provided in normalized (lower-case, snake_case) notation.
+Configuration value strings MUST be printed in quotes (so accidental leading or trailing whitespace can be spotted).
+
+Agents MUST provide information about following essential configuration items:
+
+| Item | Needs masking | Example |
+| - | - | - | - |
+| `server_url` | no | `http://localhost:8200` |
+| `secret_token` | yes | `*****` |
+| `service_name` | no | `foo` |
+| `log_level` | no | `warning` |
+
+Additional configurations items MAY be logged after that as well.
```

Comment:
The node.js agent is redacting the following config fields:

```
    const REDACT_FIELDS = {
      apiKey: true,
      secretToken: true,
      serverUrl: true,
    }
```

The `server_url` was identified as possible PII ... I think back when Kibana was adding the Node.js APM agent.

I think this part of the spec should definitely require that any agent emitting the `api_key` MUST redact that value. I suppose the `server_url` is debatable.

Reply:
The `server_url` being redacted seems strange to me tbh. It's actually a point of error and valuable information to have in the log. Also, agents (e.g. .NET) report it already.

Agreed, `api_key` (if present) should be reported and redacted by default.

Reply:
Yah, I don't know about `server_url`. The Kibana integrators considered it possible PII. I agree that it is useful for debugging/support.  Having that URL does given some information: what cloud and region they are deploying to, the name the user gave the cluster, the cluster id.

```
% curl -i https://my-deployment-31a70c.apm.us-west2.gcp.elastic-cloud.com/
HTTP/2 200
date: Tue, 15 Nov 2022 17:21:49 GMT
x-cloud-request-id: rDlozSYhRJyy-OdXUZGxIQ
x-content-type-options: nosniff
x-found-handling-cluster: cdfdba50f7a44a32981916b0faf9a7a2
x-found-handling-instance: instance-0000000000
content-length: 0
```

Reply:
I think for `server_url` the benefits outweigh the potential concerns.

a) Misconfigured server URLs are a regular source of error and confusion 
b) Server URLs are contained in (some) agent logs already. In fact, it might require actual effort to sanitize those from logs as those appear e.g. in exception text.

Thoughts though @gregkalapos and reviewers?

Reply:
Thanks @gregkalapos, good point. I'll add a note about that as well.